### PR TITLE
Detect external modification of scenes

### LINKED ISSUE #11803: If a scene is open in godot and the scene file is modified outside godot a prompt should be given on whether or not to update or overwrite
**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
Linux Mint.
Godot 2.1.4.
GPU irrelevant.

**Issue description:**
If a scene is open in Godot and the scene file is externally modified no prompt is given of this conflict, if changes are made in Godot, saving will overwrite the external modifications to the scene file.
It should prompt the user whether or not to update the scene according to the new scene file or whether to overwrite these.
Ideally it should prompt immediately as the external changes are made.

**Steps to reproduce:**
Create a scene, make a change and save.
With the scene still open in Godot then externally modify the scene file(i.e by git pulling a commit with modifications to the scene)

**Link to minimal example project:**
An example project with its own readme.
https://github.com/AlekhineDev/godot-bug-scene-conflict

### LINKED ISSUE #20250: Godot can override open scenes or scripts changed on disk by an external process (e.g. git pull)
**Godot version:**
3.0.5

**OS/device including version:**
Windows + Linux (shared)

**Issue description:**
When files are replaced in the tree, while a scene is open but has no changes pending (such as a git pull), changes are lost as the engine will try to overwrite the changes in git unless you are to close the scenes and re-open them, EXCEPT if it's the scene that is currently visible.
This also applies to changes to project.godot, as well as export templates

**Steps to reproduce:**
1. Create 2 scenes, keep both open throughout this
	- You can additionally edit the project settings and add an export template to see the other bugs
2. Save both
3. Push it to git/copy it/etc
4. Edit it elsewhere/in another godot instance
	- You can additionally edit the project settings and add/edit an export template
5. Copy back
6. Find that scenes do not correctly refresh unless you close/reopen the scenes - saving or running will destroy changes in all open scenes

Note: Other resources, and closed embedded scenes will update properly, though the latter will not visually refresh unless sub-resources that are not .tscn change.

**Minimal reproduction project:**
Not applicable

### LINKED ISSUE #22450: scene file can not refresh when edited outside of Godot Editor
**Godot version: 3.x

**OS/device including version:**

With the project opened in Godot, edit the scene file outside of the editor, the change will not reload into Godot.

**Issue description:**

**Steps to reproduce:**

**Minimal reproduction project:**

### PR DESCRIPTION
**Pull request by KoBeWi**
With this change, editor will look for external modifications in opened scenes when focused and asks for reload. This is analogical to scripts in ScriptEditorPlugin.

Closes #11803
Also closes #20250, which is a duplicate
Also closes #22450, which is a duplicate duplicate

### CODE REVIEW

**Comment by KoBeWi** [comment]
I finally got around to fixing this and the crash is gone (not sure if related to what I did xd). Also the reloading is more seamless, there's no undo history noise at all.

Probably needs some testing, but otherwise, the PR complete. I'd really like it for 3.2 >_>

**Comment by akien-mga** [comment]
We're a bit too close to 3.2-stable for such a change IMO, so I'd prefer to merge after the release. We can cherry-pick for 3.2 though if it proves to work well in the master branch.

**Comment by alexzheng** [comment]
Why such a nice feature still not available in the latest release?
Scene can not be reload automatically will overwrite the external modification.

File: editor/editor_node.cpp [hunk 1]
```diff
@@ -6483,6 +6556,28 @@ EditorNode::EditorNode() {
 
 	file_server = memnew(EditorFileServer);
 
+	disk_changed = memnew(ConfirmationDialog);
+	{
+		VBoxContainer *vbc = memnew(VBoxContainer);
+		disk_changed->add_child(vbc);
+
+		Label *dl = memnew(Label);
+		dl->set_text(TTR("The following files are newer on disk.\nWhat action should be taken?:"));
```
Comment (akien-mga) [inline editor/editor_node.cpp:6565]: `?:` is invalid punctuation, remove `:`.

File: editor/editor_node.cpp [hunk 2]
```diff
@@ -592,6 +593,72 @@ void EditorNode::_sources_changed(bool p_exist) {
...
+	}
+}
+
+void EditorNode::_resave_scenes(String p_str) {
+
+	save_all_scenes();
+	disk_changed->hide();
+}
+
+void EditorNode::_reload_scenes() {
```
Comment (akien-mga) [inline editor/editor_node.cpp:631]: I'd name it `reload_modified_scenes()` since it doesn't reload non-modified scenes a priori.

**Comment by KoBeWi** [comment]
Ok, rebased and fixed. However due to #37531, the dialog will pop up infinitely when you try to cancel/close it (because editor gets refocused), so you are either forced to save scene or reload.

**Comment by Dragoncraft89** [comment]
As far as I can tell from the diff, this only takes into account open scenes, right?

But these are not the only files, that need to be reloaded. The `project.godot` file faces similar problems, although fixing that could be a lot more difficult

**Comment by Zylann** [comment]
Note that the user can have more than just scenes open. It's actually about *resources*. Resources can be scripts, scenes, or standalone resources open in the inspector. Is the latter handled as well?

**Comment by KoBeWi** [comment]
Resources in Godot are usually built-in, so this PR handles most of the problems. But I didn't test file Resources, they might indeed be a problem. project.godot is not handled yet either.

**Comment by Zylann** [comment]
> Resources in Godot are usually built-in

No? I dunno what that comes from (we dont have stats?) but there are a bunch of cases where resources are files. I also use lots of file resources myself, and actually rarely make them builtin.

**Comment by KoBeWi** [comment]
Dunno, 99% of non-imported resources in my project are built-in (they are mostly collision shapes). Script changes are already handled by script editor.

**Comment by KoBeWi** [comment]
I rebased the PR. Also I added detecting changes to `project.godot`, but right now it's useless due to #45243 (I tested with the mentioned lines commented out and it works)

**Comment by KoBeWi** [comment]
Ah, I forgot to write here, but I tested this PR during the GGJ and it works fine.

There are two problems though:
- #37531
- built-in scripts aren't properly reloaded when the containing scene is instanced in another opened scene

However they are separate issues not really related to this PR. I'd say it's ready to be merged.

File: core/config/project_settings.cpp [hunk 3]
```diff
@@ -676,7 +677,12 @@ void ProjectSettings::clear(const String &p_name) {
 }
 
 Error ProjectSettings::save() {
-	return save_custom(get_resource_path().plus_file("project.godot"));
+	print_line("save");
```
Comment (akien-mga) [inline core/config/project_settings.cpp:680]: That's going to get annoying pretty fast :)

File: core/config/project_settings.h [hunk 4]
```diff
@@ -113,7 +115,6 @@ class ProjectSettings : public Object {
 
 	Error _setup(const String &p_path, const String &p_main_pack, bool p_upwards = false);
 
-protected:
```
Comment (akien-mga) [inline core/config/project_settings.h:116]: Why? `_bind_methods()` is always `protected`.

File: editor/editor_data.h [hunk 5]
```diff
@@ -111,6 +111,7 @@ class EditorData {
 	struct EditedScene {
 		Node *root = nullptr;
 		String path;
+		uint64_t file_modified_time;
```
Comment (akien-mga) [inline editor/editor_data.h:114]:
```suggestion
		uint64_t file_modified_time = 0;
```

File: core/config/project_settings.cpp (hunk 3 above)
Comment (KoBeWi) [reply core/config/project_settings.cpp:680]: Oops

File: core/config/project_settings.h (hunk 4 above)
Comment (KoBeWi) [reply core/config/project_settings.h:116]: It's already in the `protected` block, scroll up a bit.

File: editor/editor_node.cpp [hunk 6]
```diff
@@ -880,6 +881,81 @@ void EditorNode::_sources_changed(bool p_exist) {
...
+		uint64_t date = FileAccess::get_modified_time(editor_data.get_scene_path(i));
+
+		if (date > last_date) {
+			String filename = editor_data.get_scene_path(i);
+			editor_data.set_edited_scene(i);
+			_remove_edited_scene(false);
+
+			Error err = load_scene(filename, false, false, true, false, true);
+			if (err != OK) {
+				ERR_PRINT("Failed to load scene");
```
Comment (Calinou) [inline editor/editor_node.cpp:943]: Make the error message give some context for easier debugging:

```suggestion
				ERR_PRINT(vformat("Failed to load scene: %s", filename));
```

**Review by Calinou — DISMISSED** [review]

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 3.2.4. (Tentatively to get it well tested in the next RCs - if there are regressions I might revert and re-cherry-pick for 3.2.5 instead.)

File: editor/editor_node.cpp [hunk 7]
```diff
@@ -6599,6 +6678,30 @@ EditorNode::EditorNode() {
...
+		Label *dl = memnew(Label);
+		dl->set_text(TTR("The following files are newer on disk.\nWhat action should be taken?"));
+		vbc->add_child(dl);
+
+		disk_changed_list = memnew(Tree);
+		vbc->add_child(disk_changed_list);
+		disk_changed_list->set_v_size_flags(Control::SIZE_EXPAND_FILL);
+
+		disk_changed->connect("confirmed", callable_mp(this, &EditorNode::_reload_modified_scenes));
+		disk_changed->connect("confirmed", callable_mp(this, &EditorNode::_reload_project_settings));
```
Comment (akien-mga) [inline editor/editor_node.cpp:6696]: You're connecting two callables to the same signal, bug?

Comment (KoBeWi) [reply editor/editor_node.cpp:6696]: No, this is intended. I made reloading scenes and project settings into separate methods.

**Comment by akien-mga** [comment]
If you're up for it, I'd suggest making a dedicated backport PR for `3.2`, as the cherry-picking needs some care and proper testing to make sure that going back to `_bind_methods` for callbacks, etc. works as intended.

**Pull request merged**